src/v/storage/log_reader.cc

```diff
@@ -408,23 +408,25 @@ log_reader::do_load_slice(model::timeout_clock::time_point timeout) {
             co_await _iterator.close();
             co_return log_reader::storage_t{};
         }
-        if (_last_base == _config.start_offset) {
-            set_end_of_stream();
-            co_await _iterator.close();
-            co_return log_reader::storage_t{};
-        }
-        /**
-         * We do not want to close the reader if we stopped because requested
-         * range was read. This way we make it possible to reset configuration
-         * and reuse underlying file input stream.
-         */
+
+        // We do not want to close the reader if we stopped because requested
+        // range was read. This way we make it possible to reset configuration
+        // and reuse underlying file input stream.
+
         if (
           _config.start_offset > _config.max_offset
           || _config.bytes_consumed > _config.max_bytes
           || _config.over_budget) {
             set_end_of_stream();
             co_return log_reader::storage_t{};
         }
+
+        if (_last_base == _config.start_offset) {
```

Comment:
is this the case where the read isn't making progress (maybe because it had reached the most recent data in the log)? if so, it seems like that is a form of

> we stopped because requested range was read

(from the comment above), in which case, i wonder if the iterator can remain open here too.

if so, not suggesting we change this. what i remember is all this is fragile and if we are already getting a good win from the reordering of these checks then that's great.

Reply:
> (from the comment above), in which case, i wonder if the iterator can remain open here too.

Well, the condition for keeping the iterator open is really like "we read the requested amount (not range, per se), and that amount resulted in us stopping before the end of the log". If we got to the end of the log, then by definition there can't be any useful readahead cached in the reader, so the iterator doesn't seem worth saving (though there might be some secondary benefit, e.g., avoiding closing the log file in some cases as we close logs with no references sometimes).

That said, this made me think about something else: is readahead even _safe_? If we are in the last segment, then the tip of the log is unstable, e.g., we often overwrite the last 4K block several times, with successively fewer zeros on the right, as we keep appending. We probably never read this zero/unstable region explicitly, but it could be read readahead and then we may read that stale version later?

E.g.,

t0: offsets 0-1000 are valid, 1001+ is zero data
t1: we issue read on input_stream for 0-1000, which is safe
t2: seastar decides to read 0-4000 (actually, this doesn't even require readahead, just rounding up the read), so it reads invalid data in its buffer for 1000-4000
t3: we write offsets 1001-4000 so they are valid now
t4: we issue a read for 3000 more bytes from the input_stream (1001-3000) but these are returned from the buffer so are invalid

I.e., the input stream buffering effectively make the write at t4 seem like it happened before the write at t3, so it doesn't see the data from t3.

Even if true this is probably hard to observe in practice because reads so close to the tip of the log are likely to be batch cache hits.